common/dynamicconfig/collection.go

```diff
@@ -144,6 +150,11 @@ func (c *Collection) GetPingChecks() []pingable.Check {
 			Name:    "dynamic config callbacks",
 			Timeout: 5 * time.Second,
 			Ping: func() []pingable.Pingable {
+				c.subscriptionLock.Lock()
+				defer c.subscriptionLock.Unlock()
+				if c.callbackPool == nil {
+					return nil
```

Comment:
It looks like this is more than just a subscription lock as nothing subscription-related happens here.

What all data is this protecting?

Reply:
If it didn't hold subscriptionLock, it could submit a callback to callbackPool concurrently with callbackPool.Stop() in Stop(), which means the callback might not be called, which would be a false positive deadlock detection.

Basically subscriptionLock is protecting callbackPool. I can add a comment

Reply:
Would it be better to call it `callbackLock` then? No comment needed if so

Reply:
but it does also protect subscriptions and subscriptionIdx. and I don't think splitting it into two locks makes sense either: you'd need both when you do updates, so might as well combine them.